File: Campeonato.py
class Campeonato ():
	def __init__(self, clubes, qtde1Fase, qtdeGrupos2Fase, qtde3Fase, modo1Fase, modo2Fase, modo3Fase): #modo: 1 - jogo unico, 2 - ida e volta, 0 - nao ha fase
		self.listaClubes = clubes
		self.qtdeClubesAtual = len(clubes)
		self.qtdeClubesTotal = self.qtdeClubesAtual
		
		if qtde1Fase%2 == 0 and (0 < modo1Fase <= 2):
			self.qtdeClubesPrimeiraFase = qtde1Fase
			self.modoPrimeiraFase = modo1Fase
		else:
			self.qtdeClubesPrimeiraFase = 0
			self.modoPrimeiraFase = 0
		
		if qtdeGrupos2Fase != 0 and (self.qtdeClubesTotal - self.qtdeClubesPrimeiraFase/2)%qtdeGrupos2Fase == 0 and (0 < modo2Fase <= 2):
			self.qtdeGruposSegundaFase = qtdeGrupos2Fase
			self.modoSegundaFase = modo2Fase
		else:
			self.qtdeGruposSegundaFase = 0
			self.modoSegundaFase = modo2Fase
		
		
		if self.potencia2(qtde3Fase) and 0 < modo3Fase <= 2:
			self.qtdeClubesTerceiraFase = qtde3Fase
			self.modoTerceiraFase = modo3Fase
		else:
			self.qtdeClubesTerceiraFase = 0
			self.modoTerceiraFase = 0
		
		self.rodadaAtual = 0
		
	def proxRodada(self):
		qtdeRod1Fase = self.modoPrimeiraFase
		if self.qtdeGruposSegundaFase != 0:
			qtdeRod2Fase = qtdeRod1Fase + self.modoSegundaFase*(((self.qtdeClubesTotal- self.qtdeClubesPrimeiraFase/2)/self.qtdeGruposSegundaFase)-1)
		else:
			qtdeRod2Fase = 0
		qtdeRod3Fase = qtdeRod2Fase + self.modoTerceiraFase*self.qtdeFases3Fase(self.qtdeClubesTerceiraFase)
		if self.rodadaAtual < qtdeRod1Fase:
			self.rodadaAtual += 1
			self.primeiraFase()
			
		elif self.rodadaAtual < qtdeRod2Fase:
			self.rodadaAtual += 1
			self.segundaFase()
			
		elif  self.rodadaAtual < qtdeRod3Fase:
			self.rodadaAtual += 1
			self.terceiraFase()
			
		else:
			return False
		return True
	
	def primeiraFase(self):
		print('segunda Fase')
		
	def segundaFase(self):
		print('segunda Fase')
	
	def terceiraFase(self):
		print('terceira Fase')
		
	def potencia2(self, n):
		i = 1
		while i < n:
			i = i << 1
		if i == n:
			return True
		return False
		
	def qtdeFases3Fase(self, n):
		qtde = 0
		while n >= 2:
			n = n/2
			qtde += 1
		if n == 1:
			return qtde
		return 0

File: Campeonato/Campeonato.java
package Campeonato;

import java.util.ArrayList;

import Entidades.Clube;

public class Campeonato {
	
	public final static int SEM_FASE = 0;
	public final static int JOGO_UNICO = 1;
	public final static int JOGO_IDAVOLTA = 2;
	
	private int rodadaAtual = 0;
	private int qtdeClubesAtual;
	private int qtdeClubesTotal;
	
	private int qtdeClubesPrimeiraFase;
	private int qtdeGruposSegundaFase;
	private int qtdeClubesTerceiraFase;
	
	private int modoTerceiraFase;
	private int modoSegundaFase;
	private int modoPrimeiraFase;
	
	private int qtdeRod1Fase;
	private int qtdeRod2Fase;
	private int qtdeRod3Fase;
	
	private ArrayList<Clube> listaClubes;
	
	public Campeonato(ArrayList<Clube> clubes, 
			int qtde1Fase, int qtdeGrupos2Fase,int qtde3Fase, 
			int modo1Fase, int modo2Fase, int modo3Fase ){
		
		this.listaClubes = clubes;
		this.qtdeClubesAtual = clubes.size();
		//this.qtdeClubesAtual = 16; //teste
		this.qtdeClubesTotal = this.qtdeClubesAtual;
		this.inicializar(qtde1Fase, qtdeGrupos2Fase, qtde3Fase, modo1Fase, modo2Fase, modo3Fase);
		this.calcularRodadas();
		
	}
	
	private void inicializar(
			int qtde1Fase, int qtdeGrupos2Fase,int qtde3Fase, 
			int modo1Fase, int modo2Fase, int modo3Fase){
		if (qtde1Fase%2 == 0 && (0 < modo1Fase && modo1Fase <= 2)){
			this.qtdeClubesPrimeiraFase = qtde1Fase;
			this.modoPrimeiraFase = modo1Fase;
		} else {
			this.qtdeClubesPrimeiraFase = 0;
			this.modoPrimeiraFase = SEM_FASE;
		}
		
		if (qtdeGrupos2Fase != 0 
				&& (this.qtdeClubesTotal - this.qtdeClubesPrimeiraFase/2)%qtdeGrupos2Fase == 0 
				&& (0 < modo2Fase && modo2Fase <= 2)){
			this.qtdeGruposSegundaFase = qtdeGrupos2Fase;
			this.modoSegundaFase = modo2Fase;
		} else {
			this.qtdeGruposSegundaFase = 0;
			this.modoSegundaFase = SEM_FASE;
		}
		
		if (this.potencia2(qtde3Fase) && (0 < modo3Fase && modo3Fase <= 2)){
			this.qtdeClubesTerceiraFase = qtde3Fase;
			this.modoTerceiraFase = modo3Fase;
		} else {
			this.qtdeClubesTerceiraFase = 0;
			this.modoTerceiraFase = SEM_FASE;
		}
		
		
	}
	
	private void calcularRodadas(){
		this.qtdeRod1Fase = this.modoPrimeiraFase;
		
		if (this.qtdeGruposSegundaFase != 0){
			this.qtdeRod2Fase = this.qtdeRod1Fase + this.modoSegundaFase 
					* (((this.qtdeClubesTotal - this.qtdeClubesPrimeiraFase/2)/this.qtdeGruposSegundaFase)-1);					
		} else {
			this.qtdeRod2Fase = 0;
		}
		
		this.qtdeRod3Fase = this.qtdeRod2Fase + this.modoTerceiraFase * this.qtdeFases3Fase(this.qtdeClubesTerceiraFase);
		
	}
	
	public boolean proxRodada(){
		
		if (this.rodadaAtual < this.qtdeRod1Fase){
			this.rodadaAtual += 1;
			this.primeiraFase();
		} else if (this.rodadaAtual < this.qtdeRod2Fase) {
			this.rodadaAtual += 1;
			this.segundaFase();
			
		} else if ( this.rodadaAtual < this.qtdeRod3Fase) {
			this.rodadaAtual += 1;
			this.terceiraFase();			
		} else {
			return false;
		}
		System.out.println(this.rodadaAtual);//teste
		return true;	
	}
	
	private void primeiraFase(){
		System.out.println("Primeira Fase");
	}
	
	private void segundaFase(){
		System.out.println("Segunda Fase");
	}
	
	private void terceiraFase(){
		System.out.println("Terceira Fase");
	}
	
	private boolean potencia2(int n){
		int i = 1;
		while (i < n){
			i = i << 1;
		}
		if (i == n)
			return true;
		return false;
	}
	
	private int qtdeFases3Fase(int n){
		int qtde = 0;
		while (n >= 2){
			n = n/2;
			qtde ++;
		}
		if (n == 1)
			return qtde;
		return 0;		
	}

}
